providers/implementations/signature/dsa_sig.c

```diff
@@ -331,8 +361,80 @@ static int dsa_sign(void *vpdsactx, unsigned char *sig, size_t *siglen,
     return 1;
 }
 
-static int dsa_verify(void *vpdsactx, const unsigned char *sig, size_t siglen,
-                      const unsigned char *tbs, size_t tbslen)
+static int dsa_signverify_message_update(void *vpdsactx,
+                                         const unsigned char *data,
+                                         size_t datalen)
+{
+    PROV_DSA_CTX *pdsactx = (PROV_DSA_CTX *)vpdsactx;
+
+    if (pdsactx == NULL)
+        return 0;
+
+    return EVP_DigestUpdate(pdsactx->mdctx, data, datalen);
+}
+
+static int dsa_sign_message_final(void *vpdsactx, unsigned char *sig,
+                                  size_t *siglen, size_t sigsize)
+{
+    PROV_DSA_CTX *pdsactx = (PROV_DSA_CTX *)vpdsactx;
+    unsigned char digest[EVP_MAX_MD_SIZE];
+    unsigned int dlen = 0;
+
+    if (!ossl_prov_is_running() || pdsactx == NULL)
+        return 0;
+    if (pdsactx->mdctx == NULL)
+        return 0;
+    /*
+     * If sig is NULL then we're just finding out the sig size. Other fields
+     * are ignored. Defer to dsa_sign.
+     */
+    if (sig != NULL) {
+        /*
+         * There is the possibility that some externally provided
```

Comment:
Is this comment  relevant here?

Reply:
Ah, no.  Thanks

Reply:
Oh sorry, it's actually still relevant.  This function is also called by `dsa_digest_sign_final()`, so still potentially relevant.

Reply:
I'll rephrase the comment a bit